partiql-spi/src/main/kotlin/org/partiql/spi/utils/IntervalUtils.kt

```diff
@@ -0,0 +1,159 @@
+package org.partiql.spi.utils
+
+import org.partiql.spi.types.IntervalCode
+import org.partiql.spi.types.PType
+import org.partiql.spi.value.Datum
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.math.RoundingMode
+
+internal object IntervalUtils {
+    // According to SQL1992, whether to truncate or round in the least significant
+    // field of the result is implementation-defined. We implemented as truncation for Interval as most database systems do.
+    private val INTERVAL_ROUNDING_MODE = RoundingMode.DOWN
+
+    private const val INTERVAL_DEFAULT_FRACTIONAL_PRECISION = 6
+    private const val NANO_MAX_PRECISION = 9
+    private const val MONTHS_PER_YEAR = 12L
+    private const val SECONDS_PER_MINUTE = 60L
+    private const val MINUTES_PER_HOUR = 60L
+    private const val HOURS_PER_DAY = 24L
+    private const val SECONDS_PER_HOUR = SECONDS_PER_MINUTE * MINUTES_PER_HOUR
+    private const val SECONDS_PER_DAY = SECONDS_PER_HOUR * HOURS_PER_DAY
+    private const val NANOS_PER_SECOND = 1_000_000_000L
+
+    /**
+     * Divides an interval by a numeric value.
+     *
+     * @param interval The interval type to divide
+     * @return A function that takes an interval Datum and BigDecimal divisor, returns the divided interval
+     */
+    fun intervalDivide(interval: PType): ((Datum, Number) -> Datum) {
+        return when (interval.intervalCode) {
+            IntervalCode.YEAR,
+            IntervalCode.MONTH,
+            IntervalCode.YEAR_MONTH -> { i, number ->
+                // get total months from the interval as Long Type as it can hold the interval with max precision = 9.
+                val totalMonths: Long = i.years * MONTHS_PER_YEAR + i.months
+                val totalMonthsInDouble = totalMonths / number.toDouble()
```

Comment:
I think we should only keep the `number` argument a double if it's already an approximate number (float/double). We shouldn't convert it to double if it's already an exact number (int, decimal).